keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,833 @@
+    possibility of a race between the controller registering the apiserver
+    with the resources it can serve and receiving a request for a resource
+    that is not yet available on that apiserver).
+
+### User Stories (Optional)
+
+#### Garbage Collector
+
+The garbage collector makes decisions about deleting objects when all
+referencing objects are deleted. A discovery gap / apiserver mismatch, as
+described above, could result in GC seeing a 404 and assuming an object has been
+deleted; this could result in it deleting a subsequent object that it should
+not.
+
+This proposal will cause the GC to see either the correct object or get a 503
+(which it handles safely).
+
+#### Namespace Lifecycle Controller
+
+This controller seeks to empty all objects from a namespace when it is deleted.
+Discovery failures cause NLC to be unable to tell if objects of a given resource
+are present in a namespace. It fails safe, meaning it refuses to delete the
+namespace until it can verify it is empty: this causes slowness deleteing
+namespaces that is a common source of complaint.
+
+Additionally, if the NLC knows about a resource that the apiserver it is talking
+to does not, it may incorrectly get a 404, assume a collection is empty, and
+delete the namespace too early, leaving garbage behind in etcd. This is a
+correctness problem, the garbage will reappear if a namespace of the same name
+is recreated.
+
+This proposal addresses both problems.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Cluster admins might not read the release notes and realize they should enable
+network/firewall connectivity between apiservers. In this case clients will
+recieve 503s instead of transparently being proxied. 503 is still safer than
+today's behavior.
+
+Requests will consume egress bandwidth for 2 apiservers when proxied. We can cap
+the number if needed, but upgrades aren't that frequent and few resources are
+changed on releases, so these requests should not be common. We will count them
+with a metric.
+
+TODO: security / cert stuff.
+
+## Design Details
+
+TODO: explanation of how the handler will determine a request is for a resource
+that should be proxied.
+
+TODO: explanation of how the security handshake between apiservers works.
+* What we need to fix: random processes / external users / etc should not be
+  able to proxy requests, so the receiving apiserver needs to be able to verify
+  the source apiserver.
+* generate self-signed cert on startup, put pubkey in apiserver identity lease
+  object?
+
+### Unresolved (how we will make discovery consistent)
+
+One option is routing discovery requests from old-apiservers to the new api-server,
```

Comment:
Do we care about usecase of "disabled APIs" - e.g. v1beta1 that we stop serving in releases 1.X, but given we still have apiservers in release 1.X-1 we could technically serve such requests?

Reply:
Personally I think such APIs should be served until the last apiserver serving them is removed.

Reply:
I agree, which means that "routing from old apiservers to new apiservers" doesn't work for this pattern...

Reply:
Then we just have to convince @logicalhan :)

Reply:
I don't think that's good. Eventually we will want to do actions only possible in the transitory state, like clear the collection for resources that are being removed.

Reply:
> Re-routing to the newest apiserver is tantamount to an 'instantaneous' upgrade of the apiservers

that also strikes me as problematic, actually... it means if a new API server is in the cluster, but fails discovery requests, all discovery requests to all servers fail. For some endpoints (like APIs only served by the new servers) that's unavoidable, but for things like root discovery endpoints, I don't think we should introduce a single point of failure like that

Reply:
This is easily mitigated by (1) caching discovery locally after a single successful call so that subsequent discovery requests succeed, (2) if the cache is empty then falling back to the older discovery document. This sort of design basically ensures that we always serve the most consistent discovery doc (if available/reachable) and if we haven't done so yet, then we can serve the older doc, as if the upgrade has not yet happeneed.

Reply:
Sorry, but what are the other failure modes?

Reply:
> I don't think that's good. Eventually we will want to do actions only possible in the transitory state, like clear the collection for resources that are being removed.

A secondary argument that I'd like to make is that the KEP that actually wants to act on the intermediate state (*cough* feature flags which support transitions) should implement the discovery changes necessary to for those transitions. **This particular KEP** and what it is [intended to solve](https://kubernetes.slack.com/archives/C0EG7JC6T/p1677759135659099) does not require this at all.

Reply:
Han and I worked it out offline, the KEP is updated